### PR TITLE
Added structural nodes gamma & match to Mlir/Jlm conversion

### PR DESCRIPTION
**Pull request by ProgrammingLouis**

### CODE REVIEW

**Comment by ProgrammingLouis** [comment]
I'm working on fixing this, I don't have any errors localy which doesn't help to figure out what's wrong here.

**Review by sjalander — CHANGES_REQUESTED** [review]

File: jlm/mlir/frontend/MlirToJlmConverter.cpp
Comment (phate) [inline jlm/mlir/frontend/MlirToJlmConverter.cpp:166]: Why is that here? How can it be that a binary node has more than two operands?

Also, just skipping the conversion for a node is not really an option, is it? It also means its results are not defined for the next operation etc. This would ultimately mean that the entire conversion is failing, no?

File: jlm/mlir/frontend/MlirToJlmConverter.cpp
Comment (phate) [inline jlm/mlir/frontend/MlirToJlmConverter.cpp:280]: `auto convertedNode =`

File: jlm/mlir/frontend/MlirToJlmConverter.cpp
Comment (phate) [inline jlm/mlir/frontend/MlirToJlmConverter.cpp:276]: `//`

File: jlm/mlir/frontend/MlirToJlmConverter.cpp
Comment (phate) [inline jlm/mlir/frontend/MlirToJlmConverter.cpp:279]: Is this comment still up to date?

File: jlm/mlir/frontend/MlirToJlmConverter.cpp
Comment (phate) [inline jlm/mlir/frontend/MlirToJlmConverter.cpp:283]: So you use the check for two inputs whether it is a bit binary or not? That sounds a little bit off.

File: jlm/mlir/frontend/MlirToJlmConverter.cpp
Comment (phate) [inline jlm/mlir/frontend/MlirToJlmConverter.cpp:284]: `//`

File: jlm/mlir/frontend/MlirToJlmConverter.cpp
Comment (phate) [inline jlm/mlir/frontend/MlirToJlmConverter.cpp:296]: {inputs.begin(), inputs.end()}

Comment (phate) [inline jlm/mlir/frontend/MlirToJlmConverter.cpp:296]: Use `llvm::zext_op::Create(...)`

File: jlm/mlir/frontend/MlirToJlmConverter.cpp
Comment (phate) [inline jlm/mlir/frontend/MlirToJlmConverter.cpp:329]: `//`

File: jlm/mlir/frontend/MlirToJlmConverter.cpp
Comment (phate) [inline jlm/mlir/frontend/MlirToJlmConverter.cpp:340]: `auto rvsdgGammaNode = `

File: jlm/mlir/frontend/MlirToJlmConverter.cpp
Comment (phate) [inline jlm/mlir/frontend/MlirToJlmConverter.cpp:330]: `mlirCtlConst`

File: jlm/mlir/frontend/MlirToJlmConverter.cpp
Comment (phate) [inline jlm/mlir/frontend/MlirToJlmConverter.cpp:338]: `mlirGammaNode`

File: jlm/mlir/frontend/MlirToJlmConverter.cpp
Comment (phate) [inline jlm/mlir/frontend/MlirToJlmConverter.cpp:373]: `mlirMatch`

File: tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp
Comment (phate) [inline tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp:374]: Fix `TODO`

File: tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp
Comment (phate) [inline tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp:387]: Why do you have this compound block here?

File: tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp
Comment (phate) [inline tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp:407]: `..., {{4, 0}, {5, 1}, {6, 1}}, 2 ,2`

File: tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp
Comment (phate) [inline tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp:471]: `//`

File: tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp
Comment (phate) [inline tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp:481]: Fix `TODO`

File: tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp
Comment (phate) [inline tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp:494]: Why the block?

File: tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp
Comment (phate) [inline tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp:501]: `auto rvsdgGammaNode`

File: tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp
Comment (phate) [inline tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp:459]: `assert(matchRuleAttr.getValues().empty())`

File: tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp
Comment (phate) [inline tests/jlm/mlir/backend/TestJlmToMlirConverter.cpp:526]: What about this `TODO`?

**Review by phate — CHANGES_REQUESTED** [review]

File: jlm/mlir/backend/JlmToMlirConverter.hpp
Comment (phate) [reply jlm/mlir/backend/JlmToMlirConverter.hpp:98]: and reference

File: jlm/mlir/frontend/MlirToJlmConverter.cpp
Comment (phate) [reply jlm/mlir/frontend/MlirToJlmConverter.cpp:283]: See comment above.

File: jlm/mlir/backend/JlmToMlirConverter.cpp
Comment (ProgrammingLouis) [reply jlm/mlir/backend/JlmToMlirConverter.cpp:49]: Yes, this is noise from debugging

File: jlm/mlir/backend/JlmToMlirConverter.cpp
Comment (ProgrammingLouis) [reply jlm/mlir/backend/JlmToMlirConverter.cpp:125]: Why use auto instead of specifying the type ?

File: jlm/mlir/backend/JlmToMlirConverter.cpp
Comment (ProgrammingLouis) [reply jlm/mlir/backend/JlmToMlirConverter.cpp:329]: I'm using the "#region folding for VS Code" extension to be able to fold these regions.
Unfortunatly this extension only works with /* comments

File: jlm/mlir/backend/JlmToMlirConverter.cpp
Comment (ProgrammingLouis) [reply jlm/mlir/backend/JlmToMlirConverter.cpp:340]: See comment above

File: jlm/mlir/backend/JlmToMlirConverter.cpp
Comment (ProgrammingLouis) [reply jlm/mlir/backend/JlmToMlirConverter.cpp:359]: See comment above

File: jlm/mlir/backend/JlmToMlirConverter.cpp
Comment (ProgrammingLouis) [reply jlm/mlir/backend/JlmToMlirConverter.cpp:367]: See comment above

File: jlm/mlir/backend/JlmToMlirConverter.cpp
Comment (ProgrammingLouis) [reply jlm/mlir/backend/JlmToMlirConverter.cpp:450]: As before, why use auto ?

File: jlm/mlir/backend/JlmToMlirConverter.cpp
Comment (phate) [reply jlm/mlir/backend/JlmToMlirConverter.cpp]: 1. More readable. If I need the type, then the IDE can tell me.
2. Leads to less line changes when the type is renamed, changes namespace, refactored, etc.
3. Shorter lines, leading to fewer line breaks (readability).

File: jlm/mlir/backend/JlmToMlirConverter.cpp
Comment (phate) [reply jlm/mlir/backend/JlmToMlirConverter.cpp:327]: It is not great to have IDE specific comments in the repository. I consider this noise.

File: jlm/mlir/backend/JlmToMlirConverter.cpp
Comment (phate) [reply jlm/mlir/backend/JlmToMlirConverter.cpp]: see above.

**Review by sjalander — DISMISSED** [review]

**Comment by ProgrammingLouis** [comment]
@phate 
I think I have now addressed all of your changes. I'm waiting for your approval.

**Review by phate — APPROVED** [review]

**Pull request merged**
